### PR TITLE
Better hover for choropleth traces

### PR DESCRIPTION
**Pull request by etpinard**
fixes https://github.com/plotly/plotly.js/issues/1333

and cleans up a bunch of _geo-only_ code paths and logic.

### CODE REVIEW

File: src/plot_api/plot_api.js [hunk 1]
```diff
@@ -2762,11 +2762,6 @@ function makePlotFramework(gd) {
     fullLayout._glcontainer.enter().append('div')
         .classed('gl-container', true);
 
-    fullLayout._geocontainer = fullLayout._paperdiv.selectAll('.geo-container')
-        .data([0]);
-    fullLayout._geocontainer.enter().append('div')
```
Comment (etpinard) [inline src/plot_api/plot_api.js]: An artefact from the first iteration back in July 2015 - where I found that [`Fx.lonehover`](https://github.com/plotly/plotly.js/blob/143749ecfecf676adfaa42ff03469e273c3f8cea/src/plots/cartesian/graph_interact.js#L752-L808) worked best off a plain `<div>` (fast-forward today, this not longer the case).

File: src/plot_api/plot_api.js [hunk 2]
```diff
@@ -2810,6 +2805,9 @@ function makePlotFramework(gd) {
     // single ternary layer for the whole plot
     fullLayout._ternarylayer = fullLayout._paper.append('g').classed('ternarylayer', true);
 
+    // single geo layer for the whole plot
+    fullLayout._geolayer = fullLayout._paper.append('g').classed('geolayer', true);
```
Comment (etpinard) [inline src/plot_api/plot_api.js]: Now geo lies side-by-side with other SVG subplot types :tada:

File: src/traces/choropleth/plot.js [hunk 3]
```diff
@@ -79,54 +43,20 @@ plotChoropleth.plot = function(geo, calcData, geoLayout) {
...
-            .on('mouseout', function() {
-                Fx.loneUnhover(geo.hoverContainer);
-
-                geo.graphDiv.emit('plotly_unhover', eventData);
-            })
-            .on('mousedown', function() {
-                // to simulate the 'zoomon' event
-                Fx.loneUnhover(geo.hoverContainer);
+            .on('mouseover', function(pt) {
+                geo.choroplethHoverPt = pt;
```
Comment (etpinard) [inline src/traces/choropleth/plot.js]: `pt` here is essentially the calcdata item corresponding to the feature under the mouse cursor. `geo.choroplethHoverPt` is then used in the `hoverPoints` method. 

Alternatively, we could have looped over all the features and use something like [`turf-inside`](https://github.com/Turfjs/turf-inside) which is more similar to the `hoveron: 'fills'` algorithm, but I thought this here sufficed.

File: src/traces/choropleth/hover.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        hasText = (parts.indexOf('text') !== -1),
+        hasIdAsNameLabel = !hasName && hasLocation;
+
+    var text = [];
+
+    function formatter(val) {
+        return Axes.tickText(axis, axis.c2l(val), 'hover').text;
+    }
+
+    if(hasIdAsNameLabel) pointData.nameOverride = pt.id;
```
Comment (etpinard) [inline src/traces/choropleth/hover.js]: see https://github.com/plotly/plotly.js/pull/1401/commits/ec6c1b1651a17738c0dcb289886d4f6664e74780

**Comment by etpinard** [comment]
I can't notice any unwanted differences in the new baselines in https://github.com/plotly/plotly.js/pull/1401/commits/c32d445899038b3f278445221d3017d4a89786bb

If anything, the new baselines appear to better _clip_ the geo subplots:

gifrecord_2017-02-21_132722

I wouldn't mind having another set of :eyes: taking a look at this.

**Comment by alexcjohnson** [comment]
Beautiful. Hover effects work better AND the code and svg structure is way simpler. 🌟 💯 🌟 

I'm not worried about the baseline image differences - seems like just subpixel rounding differences based on the different nesting structure. There does seem to be a little bit of overflow in the clipping, may be worthwhile taking a 🔍  to it at some point but doesn't need to be now - I don't think the effect changed with this PR, it just got a subpixel shift.

💃

**Pull request merged**